`azurerm_kubernetes_cluster`: Support for CSI arguments

CorrenSoft opened the pull request:
May solve #19235 

Added 3 optional properties and an optional block (because the disk driver has an additional parameter to specify the version).

Example of implementation:

```
resource "azurerm_kubernetes_cluster" "aks" {
   
  ...

  storage_disk_driver {
    enabled = true
    version = "V1"
  }
  storage_blob_driver         = true
  storage_file_driver         = false
  storage_snapshot_controller = false
}
```

Considering the [SDK's Object Model](https://pkg.go.dev/github.com/hashicorp/go-azure-sdk@v0.20221114.1064854/resource-manager/containerservice/2022-09-02-preview/managedclusters#ManagedClusterStorageProfile), I have two questions on my mind, and I would like to have some feedback:
1- Should all be blocks instead of properties, to allow easy expansion if in the future extra fields are added?
2- Should be all this encapsulated in a single block instead of individual components?

**Note: The docs are yet to be updated, will do so upon receiving feedback on the code.**

stevehipwell commented:
@CorrenSoft I can confirm that this PR would resolve https://github.com/hashicorp/terraform-provider-azurerm/issues/19235.

My suggestion would be to use blocks and to potentially use a primary `storage` block to wrap the other ones.

Also is there a reason why you're using `V1` instead of `v1` for the version?

the-technat commented:
Any news on this? We @swisspost would love to see this feature merged.

stephybun commented on internal/services/containers/kubernetes_cluster_resource.go:
```diff
@@ -1004,6 +1004,48 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"storage_disk_driver": {
+				Type:     pluginsdk.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &pluginsdk.Resource{
+					Schema: map[string]*pluginsdk.Schema{
+						"enabled": {
+							Type:     pluginsdk.TypeBool,
+							Optional: true,
+							Default:  true,
+						},
+						"version": {
+							Type:     pluginsdk.TypeString,
+							Optional: true,
+							Default:  "V1",
+							ValidateFunc: validation.StringInSlice([]string{
+								"V1",
+								"V2",
+							}, false),
+						},
+					},
+				},
+			},
+
+			"storage_file_driver": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+
+			"storage_snapshot_controller": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
```
These should probably be grouped into a `storage_profile` block

stephybun commented on internal/services/containers/kubernetes_cluster_resource.go:
```diff
@@ -1004,6 +1004,48 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
 				}, false),
 			},
 
+			"storage_blob_driver": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  false,
```
Can we rename this to (also after it's been nested under a `storage_profile` block)
```suggestion
			"blob_driver_enabled": {
				Type:     pluginsdk.TypeBool,
				Optional: true,
				Default:  false,
```

stephybun commented on internal/services/containers/kubernetes_cluster_resource.go:
```diff
@@ -1004,6 +1004,48 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
+				MaxItems: 1,
+				Elem: &pluginsdk.Resource{
+					Schema: map[string]*pluginsdk.Schema{
+						"enabled": {
+							Type:     pluginsdk.TypeBool,
+							Optional: true,
+							Default:  true,
```
We're moving away from exposing the `Enabled` property to users, particularly for AKS a fair amount of refactoring went in to remove this field for enabling/disabling features within the resource.

Could this be flattened into two separate properties instead
`disk_driver_enabled` and `disk_driver_version`

stephybun commented on internal/services/containers/kubernetes_cluster_resource.go:
```diff
@@ -1004,6 +1004,48 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
+				},
+			},
+
+			"storage_file_driver": {
```
Can we rename this to 
```suggestion
			"file_driver_enabled": {
```

stephybun commented on internal/services/containers/kubernetes_cluster_resource.go:
```diff
@@ -1004,6 +1004,48 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
+				Default:  true,
+			},
+
+			"storage_snapshot_controller": {
```
And this to
```suggestion
			"snapshot_controller_enabled": {
```

stephybun reviewed (changes requested):
Thanks for this PR @CorrenSoft, this is off to a great start!

As suggested by @stevehipwell, I think these would benefit from being grouped into a block called `storage_profile`. I also left a few more comments in-line regarding the property names and the property `enabled` which we don't typically expose in Terraform.

Other than that as you've mentioned the docs need to be updated, and we also need to add some tests around this.

stevehipwell commented:
@CorrenSoft I'd suggest that `v1` works better for the Terraform API version as it's the industry wide idiomatic representation of a version. @stephybun is there a convention for this provider or Terraform in general which applies to this scenario?

CorrenSoft commented:
@stevehipwell I changed the values to `v1` and `v2`. In any case, values are not case-sensitive.

@stephybun I completed the refactor, updated the docs and added a new test (creates cluster with values different from default). Let me know if anything else is needed.

Note: Still running the Acc Tests.

stephybun commented:
@stevehipwell the convention in the AzureRM provider is to use whatever casing is specified in the API spec/enums provided by Azure. Since there are none for these and the API seems to accept both upper and lower casing I would agree with you that we should go with the industry convention for representing versions.

As an aside in the [swagger they used lower case v1 in the description](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/containerservice/resource-manager/Microsoft.ContainerService/preview/2022-09-02-preview/managedClusters.json#L6625).

stevehipwell commented:
> @stevehipwell the convention in the AzureRM provider is to use whatever casing is specified in the API spec/enums provided by Azure. Since there are none for these and the API seems to accept both upper and lower casing I would agree with you that we should go with the industry wide convention for representing versions.

> 
> As an aside in the [swagger they used lower case v1 in the description](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/containerservice/resource-manager/Microsoft.ContainerService/preview/2022-09-02-preview/managedClusters.json#L6625).

They used lowercase in the docs too but we've all been caught out by mixed case sensitivity in Azure.

CorrenSoft commented on website/docs/d/kubernetes_cluster.html.markdown:
```diff
@@ -38,7 +38,7 @@ The following attributes are exported:
 * `api_server_authorized_ip_ranges` - The IP ranges to whitelist for incoming traffic to the primaries.
 
--> **NOTE:** `api_server_authorized_ip_ranges` Is currently in Preview on an opt-in basis. To use it, enable feature `APIServerSecurityPreview` for `namespace Microsoft.ContainerService`. For an example of how to enable a Preview feature, please visit [How to enable the Azure Firewall Public Preview](https://docs.microsoft.com/azure/firewall/public-preview)
+-> **NOTE:** `api_server_authorized_ip_ranges` Is currently in Preview on an opt-in basis. To use it, enable the feature `APIServerSecurityPreview` for `namespace Microsoft.ContainerService`. For an example of how to enable a Preview feature, please visit [How to enable the Azure Firewall Public Preview](https://docs.microsoft.com/azure/firewall/public-preview)
```
Done, and also removed a couple more similar notes.

CorrenSoft commented:
> I have one more question which is, are these properties actually updateable?

Yes, all of them. The only catch is if you try to use the `v2` for the disk version because a preview feature is required (already included a note in the doc).

stephybun reviewed (approved):
Great work on this, thanks @CorrenSoft. LGTM 💯
